File: September/10th September,2024/Leetcode(2807).py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from math import *
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp=head.next
        curr=head
        if head.next is None:
            return head
        while temp:
            x=gcd(temp.val,curr.val)
            p=ListNode(x,temp)
            curr.next=p
            curr=temp
            temp=temp.next
        return head

File: September/10th September,2024/Leetcode(2807).cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* curr=head;
        if(head->next==NULL){
            return head;
        }
        ListNode* temp=head->next;
        while(temp!=NULL){
            ListNode* x=new ListNode(__gcd(temp->val,curr->val),temp);
            curr->next=x;
            curr=temp;
            temp=temp->next;
        }        
        return head;
    }
};